### ISSUE TITLE
"connection has been closed by server" in pubsub mode

### ISSUE THREAD

**Issue by yoonjechoi** [aio-libs/aioredis-py#409]
Hello, 

In pubsub mode, I encountered connection have been closed by server. It is caused by exceeding client-output-buffer-limit of redis-server and it can be happen in our environment due to heavy publish operations.

DEBUG - aioredis - [connection.py:201] - Connection has been closed by server, response: None
DEBUG - aioredis - [connection.py:377] - Closing pubsub channel <Channel name:b'antenna_channel', is_pattern:False, qsize:0>

I'm trying to re-connect and re-subscribe when connection has been closed by server.
but I cannot find the way to catch ConnectionClosedError. 

```python

async def connect_and_subscribe(redis_url, channel_name, encoding)
    redis_pool = await aioredis.create_redis_pool(redis_url, encoding=encoding)
    mpsc = Receiver(loop=asyncio.get_event_loop())
    channel_list = [mpsc.channel(channel_name) for channel_name in channel_name_list]
    subscribed_channels = await redis_poll.subscribe(*channel_list)
    return redis_pool, subscribed_channels, mpsc

async def do(redis_url, channel_name, encoding='utf-8')):
    """
        Here I try to catch ConnectionClosedError and reconnect, resubscribe
    """
    redis_pool, subscribed_channels, mpsc = connect_and_subscribe(redis_url, channel_name, encoding)
               
    async for channel, message in mpsc.iter(encoding=encoding):
        try:
            process_message(message)
                              
        except ConnectionClosedError:
            # reconnect
            redis_pool, orderbook_channel_name, mpsc = await connect_and_subscribe(redis_url, channel_name, encoding)

```

ConnecionClosedError is not catched in this way.
How should I do to catch ConnecionClosedError and reconnect properly in pub/sub mode.

(I've read auto-reconnect is not supported in pub/sub mode. is it right?)

**Comment by geerk** [comment]
I have a similar issue. How should we handle a closed connection in case of PUB/SUB?
Should this code raise a `ConnectionClosedError`?
```python
  async for channel, message in mpsc.iter():
```

What is the best practice for this kind of problem?
Thank you.

**Comment by yoonjechoi** [comment]
@geerk 

To work around the problem, I replace mpsc.iter() with channel.wait_message() and channel.get().
Command Interface has the same issue. 
@popravich  I think it may be a bug of aioredis. Would you check it?~ Thank you

Check code below


```python
    async def subscribe(self, redis_url, channel_name, encoding='utf-8'):
        try:
            redis_pool = None

            while True:
                # Accidently redis connection can be closed by client buffer full problem
                # re-create(connect) redis pool through outer while loop
                if not redis_pool or redis_pool.closed:
                    redis_pool = await aioredis.create_redis_pool(redis_url, encoding=encoding)

                res = await redis_pool.subscribe(channel_name)
                channel = res[0]
               
                while await channel.wait_message():
                     message = await channel.get(encoding=constants.DEFAULT_ENCODING)
                     do_something(message)


        finally:
            # When this is cancelled and CancelledExecption raised
            if redis_pool and not redis_pool.closed:
                redis_pool.close()
                await redis_pool.wait_closed()
```

**Comment by Fogapod** [comment]
I am having similar issue. I lose connection to redis every ~300 seconds.
```
DEBUG    aioredis Connection has been closed by server, response: None
DEBUG    aioredis Connection has been closed by server, response: None
DEBUG    aioredis Closing pubsub channel <Channel name:b'rpc:run-api', is_pattern:False, qsize:0>
```
This happens constantly on 1 of my machines and only inside docker. Running the same code outside of docker/inside docker on different machine I don't get any disconnects (the same redis server is used).

I checked debug redis logs and all I've got was this:
```
6535:M 14 Nov 08:53:09.613 - Client closed connection
6535:M 14 Nov 08:53:09.614 - Client closed connection
```

Changing tcp-timeout setting on redis server did not take any effect.

I also tried replacing `channel.iter()` with `channel.wait()` and `channel.get()` as suggested above, but It did not change anything.

The problem happenes using low level connection and command connection.

Running `redis-cli` in subscribe mode from other container on that server stops disconnects on both containers.

Is this an aioredis bug?
